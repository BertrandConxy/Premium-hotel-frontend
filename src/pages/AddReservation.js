import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faBars, faArrowLeft } from '@fortawesome/free-solid-svg-icons';
import { Formik } from 'formik';
import { Form, Button, Offcanvas } from 'react-bootstrap';
import * as Yup from 'yup';
import { useNavigate, useParams } from 'react-router-dom';
import { createReservation } from '../redux/reservations/reservations';
import NavPanel from '../components/NavPanel';
import lunar from '../images/lunar.png';
import './addReservation.css';
import { getRooms } from '../redux/rooms/rooms';

const validationSchema = Yup.object().shape({

  check_in: Yup.date()
    .default(() => new Date())
    .required('*Check In date is required'),
  check_out: Yup.date()
    .when('check_in', (checkIn, yup) => checkIn && yup.min(checkIn, 'Check in date cannot be greater than check out date'))
    .required('*Check Out date is required'),
});

const AddReservation = () => {
  const [show, setShow] = useState(false);
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const { id } = useParams();
  const rooms = useSelector((state) => state.roomsReducer);

  useEffect(() => {
    dispatch(getRooms());
  }, [dispatch]);

  const handleClose = () => setShow(false);
  const handleShow = () => setShow(true);
  const handleBack = () => navigate(-1);
  return (
    <div className="form-container1">
      <button type="button" className="p-2 btn">
        <FontAwesomeIcon
          icon={faArrowLeft}
          onClick={handleBack}
          className="text-white point"
        />
      </button>
      <div className="fullScreen">
        <button type="button" className="p-2 vis btn">
          <FontAwesomeIcon
            icon={faBars}
            onClick={handleShow}
            className="text-white"
          />
        </button>
        <div className="txtWrapper">
          <Formik
            initialValues={{
              check_in: '',
              check_out: '',
              room_id: id,
            }}
            validationSchema={validationSchema}
            onSubmit={(values, { setSubmitting, resetForm }) => {
              setSubmitting(true);
              dispatch(createReservation(values));
              resetForm();
              setSubmitting(false);
              setTimeout(() => {
                navigate('/reservations');
                window.location.reload(true);
              }, 1000);
            }}
          >
            {({
              values,
              errors,
              touched,
              handleChange,
              handleBlur,
              handleSubmit,
              isSubmitting,
            }) => (
              <Form
                onSubmit={handleSubmit}
                className="mx-auto row row-cols-2 mt-3 d-flex justify-content-center align-items-center"
              >
                <Form.Group controlId="FormCheckIn">
                  <Form.Label>Check In</Form.Label>
                  <Form.Control
                    type="date"
                    name="check_in"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.check_in}
                    className="{touched.check_in && errors.check_in ? 'error' : null} form-radius"
                  />
                  {touched.check_in && errors.check_in ? (
                    <div className="error-message-white">{errors.check_in}</div>
                  ) : null}
                </Form.Group>

                <Form.Group controlId="FormCheckIn">
                  <Form.Label>Check Out</Form.Label>
                  <Form.Control
                    type="date"
                    name="check_out"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.check_out}
                    className="{touched.check_out && errors.check_out ? 'error' : null} form-radius"
                  />
                  {touched.check_out && errors.check_out ? (
                    <div className="error-message-white">
                      {errors.check_out}
                    </div>
                  ) : null}
                </Form.Group>

                <Form.Group controlId="formBasicCity">
                  <Form.Label>Room</Form.Label>
                  {id ? (
                    <Form.Select
                      aria-label="Select Room Field"
                      name="room"
                      onChange={handleChange}
                      onBlur={handleBlur}
                      value={values.room_id}
                      disabled
                      className="disabled"
                    >
                      {Array.isArray(rooms)
                      && rooms.filter((room) => room.id === id) && rooms.map((room) => (
                        <option key={room.id} value={room.id}>
                          {room.name}
                        </option>
                      ))}
                    </Form.Select>
                  ) : (

                    <Form.Select
                      aria-label="Select Room Field"
                      name="room"
                      onChange={handleChange}
                      onBlur={handleBlur}
                      value={values.room}
                      className={touched.room && errors.room ? 'error' : null}
                    >
                      <option>Select Room</option>
                      {Array.isArray(rooms)
                      && rooms.map((room) => (
                        <option key={room.id} value={room.id}>
                          {room.name}
                        </option>
                      ))}
                    </Form.Select>
                  ) }
                  {touched.room && errors.room ? (
                    <div className="error-message">{errors.room}</div>
                  ) : null}
                </Form.Group>

                <Button
                  type="submit"
                  disabled={isSubmitting}
                  className="upperCase resBtn"
                >
                  Reserve
                </Button>
              </Form>
            )}
          </Formik>
        </div>
      </div>
      <Offcanvas className="darkened-off" show={show} onHide={handleClose}>
        <Offcanvas.Header closeButton>
          <Offcanvas.Title>
            <img src={lunar} className="lunar-logo" alt="Premium Hotel Logo" />
          </Offcanvas.Title>
        </Offcanvas.Header>
        <Offcanvas.Body>
          <NavPanel className="text-black" />
        </Offcanvas.Body>
      </Offcanvas>
    </div>
  );
};

export default AddReservation;
