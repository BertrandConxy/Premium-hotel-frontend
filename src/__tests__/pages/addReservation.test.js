import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import { screen, render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import AddReservation from '../../pages/AddReservation';
import store from '../../redux/configureStore';

const AddReservationProvider = () => (
  <Provider store={store}>
    <Router>
      <AddReservation />
    </Router>
  </Provider>
);

describe('render component', () => {
  it('renders the rooms list form', () => {
    render(<AddReservationProvider />);
    expect(screen.getByLabelText(/check in/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/check out/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/room/i)).toBeInTheDocument();
    expect(screen.getByText(/reserve/i)).toBeInTheDocument();
  });
});

describe('Snapshot test', () => {
  it('looks the same as the snapshot', () => {
    const component = renderer.create(<AddReservationProvider />).toJSON();
    expect(component).toMatchSnapshot();
  });
});
