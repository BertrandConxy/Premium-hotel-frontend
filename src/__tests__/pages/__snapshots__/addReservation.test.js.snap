// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test looks the same as the snapshot 1`] = `
<div
  className="form-container1"
>
  <button
    className="p-2 btn"
    type="button"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-arrow-left text-white point"
      data-icon="arrow-left"
      data-prefix="fas"
      focusable="false"
      onClick={[Function]}
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M447.1 256C447.1 273.7 433.7 288 416 288H109.3l105.4 105.4c12.5 12.5 12.5 32.75 0 45.25C208.4 444.9 200.2 448 192 448s-16.38-3.125-22.62-9.375l-160-160c-12.5-12.5-12.5-32.75 0-45.25l160-160c12.5-12.5 32.75-12.5 45.25 0s12.5 32.75 0 45.25L109.3 224H416C433.7 224 447.1 238.3 447.1 256z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </button>
  <div
    className="fullScreen"
  >
    <button
      className="p-2 vis btn"
      type="button"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-bars text-white"
        data-icon="bars"
        data-prefix="fas"
        focusable="false"
        onClick={[Function]}
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 96C0 78.33 14.33 64 32 64H416C433.7 64 448 78.33 448 96C448 113.7 433.7 128 416 128H32C14.33 128 0 113.7 0 96zM0 256C0 238.3 14.33 224 32 224H416C433.7 224 448 238.3 448 256C448 273.7 433.7 288 416 288H32C14.33 288 0 273.7 0 256zM416 448H32C14.33 448 0 433.7 0 416C0 398.3 14.33 384 32 384H416C433.7 384 448 398.3 448 416C448 433.7 433.7 448 416 448z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </button>
    <div
      className="txtWrapper"
    >
      <form
        className="mx-auto row row-cols-2 mt-3 d-flex justify-content-center align-items-center"
        onSubmit={[Function]}
      >
        <div>
          <label
            className="form-label"
            htmlFor="FormCheckIn"
          >
            Check In
          </label>
          <input
            className="{touched.check_in && errors.check_in ? 'error' : null} form-radius form-control"
            id="FormCheckIn"
            name="check_in"
            onBlur={[Function]}
            onChange={[Function]}
            type="date"
            value=""
          />
        </div>
        <div>
          <label
            className="form-label"
            htmlFor="FormCheckIn"
          >
            Check Out
          </label>
          <input
            className="{touched.check_out && errors.check_out ? 'error' : null} form-radius form-control"
            id="FormCheckIn"
            name="check_out"
            onBlur={[Function]}
            onChange={[Function]}
            type="date"
            value=""
          />
        </div>
        <div>
          <label
            className="form-label"
            htmlFor="formBasicCity"
          >
            Room
          </label>
          <select
            aria-label="Select Room Field"
            className="form-select"
            id="formBasicCity"
            name="room_id"
            onBlur={[Function]}
            onChange={[Function]}
            value=""
          >
            <option>
              Select Room
            </option>
          </select>
        </div>
        <button
          className="upperCase resBtn btn btn-primary"
          disabled={false}
          type="submit"
        >
          Reserve
        </button>
      </form>
    </div>
  </div>
</div>
`;
